### PR TITLE
redis command splitter refactor

### PR DESCRIPTION
**Pull request by danielhochman**
This is a refactor to move logic from the handler into the request. This will make it possible to create an abstract base class for fragmented commands (`MGET` + `MSET`, `DEL`, `EXISTS`, `TOUCH`, etc) to share some of the splitting logic.

### CODE REVIEW

File: source/common/redis/command_splitter_impl.cc [hunk 1]
```diff
@@ -21,43 +21,50 @@ RespValuePtr Utility::makeError(const std::string& error) {
...
-  std::unique_ptr<SplitRequestImpl> request_handle(new SplitRequestImpl(callbacks));
-  request_handle->handle_ =
-      conn_pool_.makeRequest(request.asArray()[1].asString(), request, *request_handle);
-  if (!request_handle->handle_) {
-    callbacks.onResponse(Utility::makeError("no upstream host"));
+SplitRequestPtr SimpleRequest::create(ConnPool::Instance& conn_pool,
+                                      const RespValue& incoming_request,
+                                      SplitCallbacks& callbacks) {
+  std::unique_ptr<SimpleRequest> request_ptr =
+      std::unique_ptr<SimpleRequest>{new SimpleRequest(callbacks)};
```
Comment (dnoe) [inline source/common/redis/command_splitter_impl.cc]: Why not `std::unique_ptr<SimpleRequest> request_ptr{new SimpleRequest(callbacks)};` ?

File: source/common/redis/command_splitter_impl.cc [hunk 2]
```diff
@@ -21,43 +21,50 @@ RespValuePtr Utility::makeError(const std::string& error) {
...
 }
 
-SplitRequestPtr MGETCommandHandler::startRequest(const RespValue& request,
-                                                 SplitCallbacks& callbacks) {
-  std::unique_ptr<SplitRequestImpl> request_handle(
-      new SplitRequestImpl(callbacks, request.asArray().size() - 1));
+SplitRequestPtr MGETRequest::create(ConnPool::Instance& conn_pool,
+                                    const RespValue& incoming_request, SplitCallbacks& callbacks) {
+  std::unique_ptr<MGETRequest> request_ptr =
+      std::unique_ptr<MGETRequest>{new MGETRequest(callbacks)};
```
Comment (dnoe) [inline source/common/redis/command_splitter_impl.cc]: Here too, can you just do:

`std::unique_ptr<MGETRequest> request_ptr{new MGETRequest(callbacks)}`

File: source/common/redis/command_splitter_impl.cc [hunk 3]
```diff
@@ -21,43 +21,50 @@ RespValuePtr Utility::makeError(const std::string& error) {
...
+  if (!request_ptr->handle_) {
+    request_ptr->callbacks_.onResponse(Utility::makeError("no upstream host"));
     return nullptr;
   }
 
-  return std::move(request_handle);
+  return std::move(request_ptr);
 }
 
-AllParamsToOneServerCommandHandler::SplitRequestImpl::~SplitRequestImpl() { ASSERT(!handle_); }
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.cc]: This ASSERT should remain. I would add back the destructor and the assert.

File: source/common/redis/command_splitter_impl.cc [hunk 4]
```diff
@@ -66,50 +73,22 @@ SplitRequestPtr MGETCommandHandler::startRequest(const RespValue& request,
...
-                                                       uint32_t num_responses)
-    : callbacks_(callbacks), pending_responses_(num_responses) {
-  pending_response_.reset(new RespValue());
-  pending_response_->type(RespType::Array);
-  std::vector<RespValue> responses(num_responses);
-  pending_response_->asArray().swap(responses);
-  pending_requests_.reserve(num_responses);
-}
-
-MGETCommandHandler::SplitRequestImpl::~SplitRequestImpl() {
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.cc]: Same. Retain destructor and all ASSERTS (if there are others that I am missing).

File: source/common/redis/command_splitter_impl.cc [hunk 5]
```diff
@@ -130,25 +109,34 @@ void MGETCommandHandler::SplitRequestImpl::onResponse(RespValuePtr&& value, uint
     break;
   }
 
-  ASSERT(pending_responses_ > 0);
-  if (--pending_responses_ == 0) {
-    ENVOY_LOG(debug, "redis: response: '{}'", pending_response_->toString());
+  ASSERT(num_pending_responses_ > 0);
+  if (--num_pending_responses_ == 0) {
+    // ENVOY_LOG(debug, "redis: response: '{}'", pending_response_->toString());
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.cc]: Why is this commented out?

File: source/common/redis/command_splitter_impl.h [hunk 6]
```diff
@@ -34,72 +34,67 @@ class CommandHandlerBase {
...
 
-  // Redis::CommandSplitter::CommandHandler
-  SplitRequestPtr startRequest(const RespValue& request, SplitCallbacks& callbacks) override;
+  void onResponse(RespValuePtr&& response);
 
-private:
-  struct SplitRequestImpl : public SplitRequest, public ConnPool::PoolCallbacks {
-    SplitRequestImpl(SplitCallbacks& callbacks) : callbacks_(callbacks) {}
-    ~SplitRequestImpl();
+  void onFailure();
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.h]: All overrides should have "override" postfix, del newline between definitions, and add // Redis::ConnPool::PoolCallbacks between them. Same in other places.

File: source/common/redis/command_splitter_impl.h [hunk 7]
```diff
@@ -34,72 +34,67 @@ class CommandHandlerBase {
...
-    ConnPool::PoolRequest* handle_{};
-  };
+  SplitCallbacks& callbacks_;
+  ConnPool::PoolRequest* handle_{};
 };
 
-class MGETCommandHandler : public CommandHandler,
-                           CommandHandlerBase,
-                           Logger::Loggable<Logger::Id::redis> {
+class MGETRequest : public SplitRequest {
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.h]: Add class comment

File: source/common/redis/command_splitter_impl.h [hunk 8]
```diff
@@ -34,72 +34,67 @@ class CommandHandlerBase {
...
+    ConnPool::PoolRequest* handle_{};
   };
+
+  SplitCallbacks& callbacks_;
+  RespValuePtr pending_response_;
+  std::vector<PendingRequest> pending_requests_;
+  uint32_t num_pending_responses_;
+};
+
+template <class RequestClass>
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.h]: Add class comment

File: source/common/redis/command_splitter_impl.cc (hunk 4 above)
Comment (danielhochman) [inline source/common/redis/command_splitter_impl.cc]: fixed both cases

File: source/common/redis/command_splitter_impl.h [hunk 9]
```diff
@@ -34,72 +34,81 @@ class CommandHandlerBase {
...
+ */
+class MGETRequest : public SplitRequest, Logger::Loggable<Logger::Id::redis> {
 public:
-  MGETCommandHandler(ConnPool::Instance& conn_pool) : CommandHandlerBase(conn_pool) {}
+  ~MGETRequest();
+
+  static SplitRequestPtr create(ConnPool::Instance& conn_pool, const RespValue& incoming_request,
+                                SplitCallbacks& callbacks);
+
+  void onChildResponse(RespValuePtr&& value, uint32_t index);
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.h]: I think these 2 methods can be private?

File: source/common/redis/command_splitter_impl.cc [hunk 10]
```diff
@@ -66,50 +81,22 @@ SplitRequestPtr MGETCommandHandler::startRequest(const RespValue& request,
...
 
-  for (uint64_t i = 1; i < request.asArray().size(); i++) {
-    request_handle->pending_requests_.emplace_back(*request_handle, i - 1);
-    SplitRequestImpl::PendingRequest& pending_request = request_handle->pending_requests_.back();
+  for (uint64_t i = 1; i < incoming_request.asArray().size(); i++) {
+    request_ptr->pending_requests_.emplace_back(*request_ptr, i - 1);
+    PendingRequest& pending_request = request_ptr->pending_requests_.back();
 
-    single_mget.asArray()[1].asString() = request.asArray()[i].asString();
-    ENVOY_LOG(debug, "redis: parallel get: '{}'", single_mget.toString());
```
Comment (mattklein123) [inline source/common/redis/command_splitter_impl.cc]: Why remove this debug print?

File: source/common/redis/command_splitter_impl.cc [hunk 11]
```diff
@@ -52,10 +53,17 @@ void SimpleRequest::cancel() {
   handle_ = nullptr;
 }
 
+MGETRequest::~MGETRequest() {
+#ifndef NDEBUG
+  for (const PendingRequest& request : pending_requests_) {
```
Comment (dnoe) [inline source/common/redis/command_splitter_impl.cc]: I'd be curious if the compiler optimizer is smart enough to remove the entire loop when ASSERT is compiled out rendering this #ifdef unnecessary.

Comment (mattklein123) [inline source/common/redis/command_splitter_impl.cc]: I kind of doubt it, but I would be curious also. :)

Comment (dnoe) [inline source/common/redis/command_splitter_impl.cc]: Prepare to be impressed. I compiled this with Clang++ -O2 -S to see the assembly (using cassert, but it should be equivalent to our macro):

```
void do_nothing_to_vector(std::vector<std::string>& foo) {
  for (const std::string& s : foo) {
    assert(s.size() > 1);
  }
}
```
If you add -DNDEBUG to the Clang command line the assembly it generates is just "retq"

Comment (mattklein123) [inline source/common/redis/command_splitter_impl.cc]: cool

**Pull request merged**